File: leetcode/p00652/p00652.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        res = []
        seen = {}

        def dfs(node: Optional[TreeNode], s1: str) -> str:
            if node.left:
                s1 = dfs(node.left, s1)
            else:
                s1 += "n"
            
            s2 = ""
            if node.right:
                s2 = dfs(node.right, s2)
            else:
                s2 += "n"
            
            s = s1 + s2 + "0" + str(node.val)
            if s not in seen:
                seen[s] = 0
            elif s in seen and seen[s] == 0:
                res.append(node)
                seen[s] = 1
            
            return s

        dfs(root, "")
        return res

if __name__ == '__main__':
    s = Solution()

    # Example 1
    l20 = TreeNode(4, None, None)
    l10 = TreeNode(2, l20, None)
    l32 = TreeNode(4, None, None)
    l22 = TreeNode(2, l32, None)
    l23 = TreeNode(4, None, None)
    l11 = TreeNode(3, l22, l23)
    root = TreeNode(1, l10, l11)

    assert s.findDuplicateSubtrees(root) == [l32, l22]

File: leetcode/p00652/p00652.go
package p00652

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(2^n), where n is the depth of tree
// Space Complexity: O(2^n)
func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
	memo := map[string]bool{}
	duplicates := map[string]*TreeNode{}

	dfs(memo, duplicates, root)

	res := []*TreeNode{}
	for _, node := range duplicates {
		res = append(res, node)
	}
	return res
}

func dfs(memo map[string]bool, duplicates map[string]*TreeNode, root *TreeNode) string {
	if root == nil {
		return "x"
	}

	key := fmt.Sprintf("%v,%v,%v",
		root.Val,
		dfs(memo, duplicates, root.Left),
		dfs(memo, duplicates, root.Right),
	)

	if memo[key] {
		duplicates[key] = root
		return key
	}

	memo[key] = true
	return key
}
